Microsoft.DotNet.Interactive.PowerShell/PowerShellKernel.cs

```diff
@@ -0,0 +1,272 @@
+            Name = DefaultKernelName;
+
+            // Add Modules directory that contains the helper modules
+            string psModulePath = Environment.GetEnvironmentVariable("PSModulePath");
+
+            Environment.SetEnvironmentVariable("PSModulePath",
+                psModulePath + Path.PathSeparator + Path.Join(
+                    Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location),
+                    "Modules"));
+        }
+
+        #region Overrides
+
+        protected override Task HandleAsync(
+            IKernelCommand command,
+            KernelInvocationContext context)
+        {
+            if (command is KernelCommandBase kb)
+            {
+                if (kb.Handler == null)
+                {
+                    switch (command)
+                    {
+                        case SubmitCode submitCode:
+                            submitCode.Handler = async (_, invocationContext) =>
+                            {
+                                await HandleSubmitCode(submitCode, context);
+                            };
+                            break;
+
+                        case RequestCompletion requestCompletion:
+                            requestCompletion.Handler = async (_, invocationContext) =>
+                            {
+                                await HandleRequestCompletion(requestCompletion, invocationContext);
+                            };
+                            break;
+
+                        case CancelCurrentCommand interruptExecution:
+                            interruptExecution.Handler = async (_, invocationContext) =>
+                            {
+                                await HandleCancelCurrentCommand(interruptExecution, invocationContext);
+                            };
+                            break;
+                    }
+                }
+            }
+
+            return Task.CompletedTask;
+        }
+
+        #endregion
+
+        #region Handlers
+
+        private async Task HandleSubmitCode(
+                SubmitCode submitCode,
+                KernelInvocationContext context)
+        {
+            CancellationTokenSource cancellationSource;
+            lock (_cancellationSourceLock)
+            {
+                cancellationSource = _cancellationSource;
+            }
+            
+            // Acknowledge that we received the request.
+            var codeSubmissionReceived = new CodeSubmissionReceived(submitCode);
+            context.Publish(codeSubmissionReceived);
+
+            // Test is the code we got is actually able to run.
+            var code = submitCode.Code;
+            if (IsCompleteSubmission(code))
+            {
+                context.Publish(new CompleteCodeSubmissionReceived(submitCode));
+            }
+            else
+            {
+                context.Publish(new IncompleteCodeSubmissionReceived(submitCode));
```

Comment:
Does `IncompleteCodeSubmissionReceived` literally means the code is incomplete? `IsCompleteSubmission` returns `true` as long as the script contains parsing error, but it could be complete. My concern is about what message will the user see in that case.

Reply:
Yes, this is the intent. This is important for languages that allow multi-line statements. Something like this (in C#) would be valid in a REPL:

```csharp
var x =
```

but not in a notebook, where we don't want statements to break across cells because it makes the current state of the notebook hard to understand.

Reply:
Thanks @jonsequitur for the clarification!
Then `IsCompleteSubmission` should explicitly check for `System.Management.Automation.IncompleteParseException`.

Reply:
@daxian-dbw does `Parser.ParseInput` throw an `IncompleteParseException` or is that exception wrapped in a `ParseError`?

Reply:
I've followed what PowerShell does (checking the first ParseError for if the input is Incomplete